### PR TITLE
FBX-135 fix incorrect skinning if character not fully in bindpose

### PR DESCRIPTION
**Pull request by vkovec**

### CODE REVIEW

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 1]
```diff
@@ -1023,19 +1023,46 @@ SkinnedMeshRenderer unitySkin
...
-        ){
+            ref SkinnedMeshBoneInfo boneInfo
+        )
+        {
+            var boneDict = boneInfo.boneDict;
+            var skinnedMesh = boneInfo.skinnedMesh;
+            var boneToBindPose = boneInfo.boneToBindPose;
+
             Matrix4x4 bindPose;
+            if(boneToBindPose.TryGetValue(unityBone, out bindPose))
```
Comment (benoithudson) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1037]: Please add comments.

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 2]
```diff
@@ -1023,19 +1023,46 @@ SkinnedMeshRenderer unitySkin
...
+            {
+                return bindPose;
+            }
+
             int index;
-            if (boneDict.TryGetValue (unityBone, out index)) {
-                bindPose = bindPoses [index];
-            } else {
-                bindPose = unityBone.worldToLocalMatrix * skinnedMesh.transform.localToWorldMatrix;
+            if (boneDict.TryGetValue(unityBone, out index))
```
Comment (benoithudson) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1043]: comment here (this is checking if it's a bone with a bind pose right?)

File: com.unity.formats.fbx/Editor/FbxExporter.cs [hunk 3]
```diff
@@ -1023,19 +1023,46 @@ SkinnedMeshRenderer unitySkin
...
+                return bindPose;
+            }
+
+            // get the bindpose of the parent
+            var parentPose = GetBindPose(unityBone.parent, bindPoses, ref boneInfo);
+            // get the local transformation matrix of the bone, then transform it into
+            // the global transformation matrix with the parent in the bind pose.
+            bindPose = parentPose.inverse * Matrix4x4.TRS(unityBone.localPosition, unityBone.localRotation, unityBone.localScale);
+            // bindpose is the inverse of the transformation matrix
+            bindPose = bindPose.inverse;
```
Comment (benoithudson) [inline com.unity.formats.fbx/Editor/FbxExporter.cs:1064]: We can simplify this and save an inverse (and thus reduce numerical noise): recall the identity `(AB)^{-1} = B^{-1}A^{-1}`

Currently you have
`bindPose = (parentPose.inverse * TRS).inverse`
which should be equal to
`bindPose = TRS.inverse * parentPose.inverse.inverse`
which is mathematically
`bindPose = TRS.inverse * parentPose`
...
...
I think.

File: com.unity.formats.fbx/Editor/FbxExporter.cs (hunk 2 above)
Comment (vkovec) [reply com.unity.formats.fbx/Editor/FbxExporter.cs:1043]: yes

**Review by benoithudson — APPROVED** [review]

**Pull request merged**